lib/wallaroo/core/topology/router.pony

```diff
@@ -896,6 +911,48 @@ class val LocalPartitionRouter[In: Any val,
       Fail()
     end
 
+  fun rebalance_steps_shrink(names_bs: Array[(String, OutgoingBoundary)],
+    state_name: String, router_registry: RouterRegistry) =>
+    let steps_to_migrate = Array[(String, OutgoingBoundary, Key, U128, Step)]
+    var i: USize = 0
+
+    for (key, target) in _partition_routes.pairs() do
+      match target
+      | let s: Step =>
+        try
+          let step_id = _step_ids(key)?
+          (let target_worker: String, let boundary: OutgoingBoundary) =
+            names_bs(i.mod(names_bs.size()))?
+          steps_to_migrate.push((target_worker, boundary, key, step_id, s))
+          i = i + 1
+        else
+          Fail()
+        end
+      end
+    end
+    rebalance_steps_common(state_name, router_registry,
+      steps_to_migrate, names_bs.size())
+
+
+  fun rebalance_steps_common(state_name: String,
+    router_registry: RouterRegistry,
+    steps_to_migrate: Array[(String, OutgoingBoundary, Key, U128, Step)],
+    num_boundaries: USize) =>
+    @printf[I32]("^^Migrating %lu steps to %d workers\n".cstring(),
+      steps_to_migrate.size(), num_boundaries)
+    for (_, _, _, step_id, _) in steps_to_migrate.values() do
+      router_registry.add_to_step_waiting_list(step_id)
```

Comment:
Even though this was in a separate for loop before, it makes sense to move it into the other loop since they're iterating over the same collection.